keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -221,31 +231,54 @@ class SqlArtifactRepository(
     }
   }
 
-  override fun versions(artifact: DeliveryArtifact, limit: Int): List<String> =
-    if (isRegistered(artifact.name, artifact.type)) {
-      sqlRetry.withRetry(READ) {
-        jooq
-          .select(ARTIFACT_VERSIONS.VERSION, ARTIFACT_VERSIONS.RELEASE_STATUS)
-          .from(ARTIFACT_VERSIONS)
-          .where(ARTIFACT_VERSIONS.NAME.eq(artifact.name))
-          .and(ARTIFACT_VERSIONS.TYPE.eq(artifact.type))
-          .apply { if (artifact.statuses.isNotEmpty()) and(ARTIFACT_VERSIONS.RELEASE_STATUS.`in`(*artifact.statuses.map { it.toString() }.toTypedArray())) }
-          .fetch()
-          .getValues(ARTIFACT_VERSIONS.VERSION)
-      }
-        .sortedWith(artifact.versioningStrategy.comparator)
-        .also { versions ->
-          // FIXME: remove special handling for Docker
-          // FIXME: limit the sql query, not the returned list (once we'll have a native sorting mechanism)
-          return if (artifact is DockerArtifact) {
-            filterDockerVersions(artifact, versions, limit)
+  override fun versions(artifact: DeliveryArtifact, limit: Int): List<String> {
+    if (!isRegistered(artifact.name, artifact.type)) {
+      throw NoSuchArtifactException(artifact)
+    }
+
+    val versions = sqlRetry.withRetry(READ) {
+      jooq
+        .select(ARTIFACT_VERSIONS.VERSION, ARTIFACT_VERSIONS.RELEASE_STATUS)
+        .from(ARTIFACT_VERSIONS)
+        .where(ARTIFACT_VERSIONS.NAME.eq(artifact.name))
+        .and(ARTIFACT_VERSIONS.TYPE.eq(artifact.type))
+        .apply {
+          if (artifact.filteredByReleaseStatus) {
+            and(ARTIFACT_VERSIONS.RELEASE_STATUS.`in`(*artifact.statuses.map { it.toString() }.toTypedArray()))
           } else {
-            versions.subList(0, Math.min(versions.size, limit))
+            if (artifact.filteredByPullRequest) {
+              and(ARTIFACT_VERSIONS_PR_NUMBER.isNotNull).and(ARTIFACT_VERSIONS_PR_NUMBER.ne(EMPTY_PR_NUMBER))
+            }
+
+            if (artifact.filteredByBranch) {
+              // TODO: should we also be comparing the repo with what's configured for the app in front50?
+              and(ARTIFACT_VERSIONS_BRANCH.likeRegex(artifact.fromBranch))
+            }
+
+            // With branches or pull requests, delegate sorting and limiting to the database
+            if (artifact.filteredByPullRequest || artifact.filteredByBranch) {
```

Comment:
This check is needed here because apparently we are passing an empty set of statuses to indicate "ignore statuses" (there are tests for that which break if this check is not here).

Reply:
I find all of these `if` conditions a bit confusing.. I worry that someone else that will see it can mess it up pretty quickly 🤷

Reply:
Any suggestions?

Reply:
not really :(